subxt/src/backend/mod.rs

```diff
@@ -336,177 +336,293 @@ pub struct StorageResponse {
-            async fn from_iter<
-                T: Serialize,
-                S: IntoIterator<Item = RpcResult<Vec<RpcResult<T>>>>,
-            >(
-                items: S,
-            ) -> Self {
-                let sub = Self::new();
-                for i in items {
-                    let i: RpcResult<Vec<Item>> = i.map(|items| {
-                        items
-                            .into_iter()
-                            .map(|item| item.map(|i| serde_json::to_string(&i).unwrap()))
-                            .collect()
-                    });
-                    sub.write(i).await
-                }
-                sub
+            pub async fn write(&self, items: Message<Item>) {
+                match items {
+                    Message::Many(items) => {
+                        for i in items.unwrap() {
+                            self.sender.send(i).await.unwrap()
+                        }
+                    }
+                    Message::Single(item) => self.sender.send(item).await.unwrap(),
+                };
+            }
+
+            pub async fn write_delayed(&self, items: Message<Item>) {
+                let sender = self.sender.clone();
+                tokio::spawn(async move {
+                    tokio::time::sleep(Duration::from_millis(500)).await;
+
+                    match items {
+                        Message::Many(items) => {
+                            for i in items.unwrap() {
+                                let _ = sender.send(i).await;
+                            }
+                        }
+                        Message::Single(item) => sender.send(item).await.unwrap(),
+                    };
+                });
             }
+        }
+
+        #[derive(Default)]
+        struct InnerMockedRpcClient {
+            data_table: MockDataTable,
+            subscription_channel: Option<Subscription>,
+            subscription_handlers: HashMap<String, SubscriptionHandler>,
+            method_handlers: HashMap<String, MethodHandler>,
+        }
 
-            async fn read(&mut self) -> RpcResult<Vec<Item>> {
-                self.receiver.recv().await.unwrap()
+        impl InnerMockedRpcClient {
+            async fn call<'a>(
+                &'a mut self,
+                method_handler: &str,
+                params: Option<Box<serde_json::value::RawValue>>,
+            ) -> RawRpcFuture<'a, Box<serde_json::value::RawValue>> {
+                let method = self.method_handlers.get(method_handler).unwrap_or_else(|| {
+                    panic!(
+                        "no method named {} registered. Params: {:?}",
+                        method_handler, params
+                    )
+                });
+
+                (*method)(&mut self.data_table, &mut self.subscription_channel, params)
             }
 
-            async fn write(&self, items: RpcResult<Vec<Item>>) {
-                self.sender.send(items).await.unwrap()
+            async fn subscribe<'a>(
```

Comment:
Is there a reason for this and `call`, above, to be async? it means that you have `.await.await` below. Can you have `fn call` and `fn subscribe` instead?

Reply:
I think the reason is because the mock data is behind an async mutex, that shouldn't be that hard to get rid of I guess because we don't hold it over other await points...

Reply:
removed `async`